### PR TITLE
Fix concurrency bug(s)

### LINKED ISSUE #77: 0.0.15 no longer creates PRs.
**Describe the bug**
The latest version, `0.0.15`, no longer creates PRs. I tried same command with `0.0.14` binary and it works fine.

**To Reproduce**
Steps to reproduce the behavior including the relevant Terraform/Terragrunt/Packer version number and any code snippets and module inputs you used.

```hcl
	./git-xargs   --branch-name mkow-test4  --repos ./java-apps.txt   --commit-message "My message"   --pull-request-title "My title"   --loglevel DEBUG   `pwd`/scripts/myScript.sh
```

**Expected behavior**
The PR should be created.

**Nice to have**
- [x] Terminal output
- [ ] Screenshots

**Additional context**

Teminal output:
```
$ git-xargs   --branch-name mkow-test2  --repos ./java-apps.txt   --commit-message "My message"   --pull-request-title "My title"   --loglevel DEBUG   ***/updateServiceInfra.sh
[git-xargs] INFO[2022-06-23T14:48:59-04:00] git-xargs running...
[git-xargs] DEBU[2022-06-23T14:48:59-04:00] Looking up filename provided repo             Name=cplus-messaging-application Organization=****
[git-xargs] DEBU[2022-06-23T14:48:59-04:00] Successfully fetched repo                     Name=cplus-messaging-application Organization=***
[git-xargs] DEBU[2022-06-23T14:48:59-04:00] Repo will have all targeted scripts run against it  Repository=cplus-messaging-application
[git-xargs] DEBU[2022-06-23T14:48:59-04:00] Attempting to clone repository using GITHUB_OAUTH_TOKEN  Repo=cplus-messaging-application
[git-xargs] DEBU[2022-06-23T14:49:00-04:00] Enumerating objects: 951, done.
Counting objects: 100% (130/130), done.
Compressing objects: 100% (78/78), done.
Total 951 (delta 44), reused 106 (delta 28), pack-reused 821  Repo=cplus-messaging-application
[git-xargs] DEBU[2022-06-23T14:49:00-04:00] Created branch                                Branch Name=refs/heads/mkow-test2 Repo=cplus-messaging-application
[git-xargs] DEBU[2022-06-23T14:49:00-04:00]                                               Repo=cplus-messaging-application
[git-xargs] DEBU[2022-06-23T14:49:01-04:00] Executing command against local clone of repo...  Command="[***updateServiceInfra.sh]" Directory=***/workspace/git-xargs-cplus-messaging-application1237588432 Repo=cplus-messaging-application
[git-xargs] DEBU[2022-06-23T14:49:31-04:00] Output of command [***/scripts/updateServiceInfra.sh] for repo cplus-messaging-application in directory ***/workspace/git-xargs-cplus-messaging-application1237588432:
[git-xargs] DEBU[2022-06-23T14:49:31-04:00] Local repository worktree no longer clean, will stage and add new files and commit changes  Repo=cplus-messaging-application
[git-xargs] DEBU[2022-06-23T14:49:32-04:00] Successfully pushed local branch to remote origin  Repo=cplus-messaging-application
[git-xargs] INFO[2022-06-23T14:49:32-04:00] Repository successfully processed             Repo name=cplus-messaging-application


*****************************************************************
  GIT-XARGS RUN SUMMARY @ 2022-06-23 18:49:32.233789 +0000 UTC
  Runtime in seconds: 32
*****************************************************************


COMMAND SUPPLIED

[/***/updateServiceInfra.sh]

REPO SELECTION METHOD USED FOR THIS RUN - (see README.md for more information)

repos-file
 REPOS SUPPLIED VIA --repos FILE FLAG
│───────────────────│─────────────────────────────│
│ ORGANIZATION NAME │ URL                         │
│───────────────────│─────────────────────────────│
│ ***      │ cplus-messaging-application │
│───────────────────│──────────────────[git-xargs] DEBU[2022-06-23T14:49:32-04:00] pullRequestWorker received pull request job. Delay: 0 seconds. Retries: 0 for repo: cplus-messaging-application on branch: refs/heads/mkow-test2
[git-xargs] DEBU[2022-06-23T14:49:32-04:00] openPullRequest received job with retries: 0. Config max retries for this run: 3
───────────│

 ALL REPOS THAT WERE TARGETED FOR PROCESSING AFTER FILTERING MISSING / MALFORMED REPOS
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ REPO NAME                   │ REPO URL                                                    │
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ cplus-messaging-application │ https://***/cplus-messaging-application │
│─────────────────────────────│─────────────────────────────────────────────────────────────│


 REPOS THAT WERE SUCCESSFULLY CLONED TO THE LOCAL FILESYSTEM
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ REPO NAME                   │ REPO URL                                                    │
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ cplus-messaging-application │ https://***/cplus-messaging-application │
│─────────────────────────────│─────────────────────────────────────────────────────────────│


 REPOS THAT SHOWED FILE CHANGES TO THEIR WORKING DIRECTORY FOLLOWING COMMAND EXECUTION
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ REPO NAME                   │ REPO URL                                                    │
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ cplus-messaging-application │ https://***/cplus-messaging-application │
│─────────────────────────────│─────────────────────────────────────────────────────────────│


 REPOS WHOSE SPECIFIED BRANCHES DID NOT EXIST ON THE REMOTE, AND SO WERE FIRST CREATED LOCALLY
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ REPO NAME                   │ REPO URL                                                    │
│─────────────────────────────│─────────────────────────────────────────────────────────────│
│ cplus-messaging-application │ https://***/cplus-messaging-application │
│─────────────────────────────│─────────────────────────────────────────────────────────────│
```
Note the async log output from the PR creation thread in the report.

### PR DESCRIPTION
**Pull request by zackproser**
These changes fix some sneaky concurrency bugs that were manifesting in various ways, either with PRs failing to be opened correctly or occasional opaque waitgroup errors that are hard to replicate because they're all timing related. 

The core issue was mishandling of the waitgroup that was being created to orchestrate the work of operating on repositories and handling pull request opening with awareness of the GitHub API rate limits. 

These changes simplify the waitgroup logic: 
- Uses the stdlib's sync.WaitGroup 
- Only call waitgroup.Add once at the beginning of processing and set its length to the total number of repositories to be processed
-  Plumb the waitgroup through the entire workflow, including the pull request retry workflow that was introduced in #67 

These changes make it much simpler to reason about the concurrent logic and in local testing they resolve the issues observed in #77 and #74. 

## Description

Fixes #77 and #74.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### Migration Guide

The `--max-concurrent-repos` flag has been removed. Please update any earlier git-xargs invocations / scripts that use that flag to remove it.

### CODE REVIEW

File: repository/process.go [hunk 1]
```diff
@@ -42,35 +44,33 @@ func openPullRequestsWithThrottling(gitxargsConfig *config.GitXargsConfig, pr ty
 func ProcessRepos(gitxargsConfig *config.GitXargsConfig, repos []*github.Repository) error {
 	logger := logging.GetLogger("git-xargs")
 
+	p, _ := pterm.DefaultProgressbar.WithTotal(len(repos)).WithTitle("Processing repos").Start()
+
 	// Limit the number of concurrent goroutines using the MaxConcurrentRepos config value
 	// MaxConcurrentRepos == 0 will fall back to unlimited (previous default behavior)
```
Comment (yorinasub17) [inline repository/process.go:50]: This comment is out of date now, right?

**Review by yorinasub17 — COMMENTED** [review]
I think the current goroutines are overengineered. I am pretty sure there is a way to cut down the number of goroutines being spawned here to make it easier to follow while also fixing the concurrency bug. See comments for more details.

File: repository/process.go [hunk 2]
```diff
@@ -42,35 +44,33 @@ func openPullRequestsWithThrottling(gitxargsConfig *config.GitXargsConfig, pr ty
 func ProcessRepos(gitxargsConfig *config.GitXargsConfig, repos []*github.Repository) error {
 	logger := logging.GetLogger("git-xargs")
 
+	p, _ := pterm.DefaultProgressbar.WithTotal(len(repos)).WithTitle("Processing repos").Start()
```
[the remark this answers is not in the document]
Comment (zackproser) [reply repository/process.go]: Done in 9d07fb3

File: repository/process.go [hunk 3]
```diff
@@ -87,51 +87,63 @@ func ProcessRepos(gitxargsConfig *config.GitXargsConfig, repos []*github.Reposit
...
-func processRepo(config *config.GitXargsConfig, repo *github.Repository) error {
+func processRepo(config *config.GitXargsConfig, repo *github.Repository, wg *sync.WaitGroup, p *pterm.ProgressbarPrinter) error {
 	logger := logging.GetLogger("git-xargs")
 
 	// Create a new temporary directory in the default temp directory of the system, but append
 	// git-xargs-<repo-name> to it so that it's easier to find when you're looking for it
 	repositoryDir, localRepository, cloneErr := cloneLocalRepository(config, repo)
 
 	if cloneErr != nil {
+		wg.Done()
```
[the remark this answers is not in the document]
Comment (zackproser) [reply repository/process.go]: Thanks, Yori! Implemented your suggestions in 428e626 - so far in local testing they're looking good!

File: repository/process.go [hunk 4]
```diff
@@ -42,35 +44,33 @@ func openPullRequestsWithThrottling(gitxargsConfig *config.GitXargsConfig, pr ty
...
-
 		go func() {
 			for {
 				select {
 				case pr := <-gitxargsConfig.PRChan:
-					wg.Add()
-					go openPullRequestsWithThrottling(gitxargsConfig, pr, &wg)
+					go openPullRequestsWithThrottling(gitxargsConfig, pr)
 				}
 			}
```
[the remark this answers is not in the document]
Comment (zackproser) [reply repository/process.go]: Makes sense - implemented a fix in 428e626

File: repository/process.go [hunk 5]
```diff
@@ -69,8 +60,11 @@ func ProcessRepos(gitxargsConfig *config.GitXargsConfig, repos []*github.Reposit
 					"Repo name": repo.GetName(), "Error": processErr,
 				}).Debug("Error encountered while processing repo")
 			}
-			return processErr
 
+			p.Increment()
+			wg.Done()
```
Comment (yorinasub17) [inline repository/process.go:65]: NIT: probably better to do a `defer` at the top, although effectively this is the same. `defer` is just more defensive coding if this logic is ever extended.

**Review by yorinasub17 — APPROVED** [review]
Updates LGTM, and definitely much easier to follow and reason about!

**Review by yorinasub17 — APPROVED** [review]

**Comment by zackproser** [comment]
Thanks for reviews! Going to merge this in now.

**Pull request merged**